### PR TITLE
Improve the CNI implementation, and documentation

### PR DESCRIPTION
**Pull request by luxas**
Fixes: #171 
Fixes: #303 

This PR:
 - Improves the network plugins interface, and makes the docker implementation consistent
 - Makes ignite support multiple initialized network plugins
 - Adds networking documentation

cc @twelho

### CODE REVIEW

File: docs/networking.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+**Pros:**
+
+- **Quick start**: If you're running docker, you can get up and running without installing extra software
+- **Port mapping support**: This mode supports port mappings from the VM to the host
+
+**Cons:**
+
+- **docker-dependent**: By design, this mode is can only be used with docker, and is hence not portable across container runtimes.
+- **No multi-node support**: The IP is local (in the `172.17.0.0/16` range), and hence other computers can't connect to your VM's IP address.
```
Comment (twelho) [inline docs/networking.md]: This IP range differs from the one given above, I believe this one here (`172.17.0.0/16`) is correct and the one above should be changed.

File: pkg/network/docker/docker.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+)
+
+const pluginName = "docker-bridge"
+
+type dockerNetworkPlugin struct {
+	runtime runtime.Interface
+}
+
+func GetDockerNetworkPlugin(r runtime.Interface) (network.Plugin, error) {
+	return &dockerNetworkPlugin{r}, nil
```
Comment (twelho) [inline pkg/network/docker/docker.go]: Does this need to return an error?

File: pkg/network/docker/docker.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+	}
+
+	return &network.Result{
+		Addresses: []network.Address{
+			{
+				IPNet: net.IPNet{
+					IP:   result.IPAddress,
+					Mask: net.IPv4Mask(255, 255, 0, 0),
+				},
+				Gateway: net.IPv4(result.IPAddress[0], result.IPAddress[1], result.IPAddress[2], 1),
```
Comment (twelho) [inline pkg/network/docker/docker.go]: Add TODO to make this configurable/auto-retrievable if someone has changed their gateway?

File: pkg/network/cni/cni.go [hunk 4]
```diff
@@ -162,22 +163,46 @@ func (plugin *cniNetworkPlugin) Status() error {
 	return plugin.checkInitialized()
 }
 
-func (plugin *cniNetworkPlugin) SetupContainerNetwork(containerid string) error {
+func (plugin *cniNetworkPlugin) PrepareContainerSpec(container *runtime.ContainerConfig) error {
+	// we will handle the networking on our own
```
Comment (twelho) [inline pkg/network/cni/cni.go]: we -> We

File: pkg/operations/start.go [hunk 5]
```diff
@@ -108,21 +110,13 @@ func StartVM(vm *api.VM, debug bool) error {
...
-		}
-
-		log.Infof("Networking is now handled by CNI")
+	// Set up the networking
+	result, err := networkPlugin.SetupContainerNetwork(containerID)
+	if err != nil {
+		return err
 	}
 
+	log.Infof("Networking is handled by %s", networkPlugin.Name())
```
Comment (twelho) [inline pkg/operations/start.go]: %s -> %q for UX consistency

File: pkg/providers/docker/runtime.go [hunk 6]
```diff
@@ -2,12 +2,23 @@ package docker
...
-	providers.Runtime, err = docker.GetDockerClient()
+	log.Trace("Initializing the Docker runtime provider...")
+	providers.Runtime, err = runtime.GetDockerClient()
 	return
 }
+
+func SetDockerNetwork() error {
+	log.Trace("Initializing the Docker network provider...")
+	plugin, err := network.GetDockerNetworkPlugin(providers.Runtime)
+	if err != nil {
```
Comment (twelho) [inline pkg/providers/docker/runtime.go]: This error is unnecessary, it's always nil

File: pkg/providers/providers.go [hunk 7]
```diff
@@ -22,6 +22,10 @@ var Storage storage.Storage
 
 type ProviderInitFunc func() error
 
+func init() {
+	NetworkPlugins = make(map[string]network.Plugin)
```
Comment (twelho) [inline pkg/providers/providers.go]: This can be an inline initialization: `var NetworkPlugins = make(map[string]network.Plugin)`, no need for `init()`

**Comment by luxas** [comment]
Comments addressed, and the CI is green, merging :+1:

**Pull request merged**